src/prototype/WingetCreateMakeDSC.ps1

```diff
@@ -1,110 +1,123 @@
-  throw [System.Net.WebException]::new("Error retrieving powershell module: Microsoft.Winget.Client. Check that you have installed the Windows Package Manager modules correctly.", $_.Exception)
-  #bugbug is this good enough?
+    # If there was an exception while installing, pass it as an InternalException for further debugging
+    throw [UnmetDependencyException]::new("'Microsoft.Winget.Client' was not installed successfully", $_.Exception)
+  } finally {
+    # Check to be sure it acutally installed
+    if (-not(Get-Module -ListAvailable -Name Microsoft.Winget.Client)) {
+      throw [UnmetDependencyException]::new("`Microsoft.Winget.Client` was not found. Check that you have installed the Windows Package Manager modules correctly.")
+    }
   }
 }
 
-if ($null -eq (Get-InstalledModule -Name powershell-yaml))
-{
+# Ensure the powershell-yaml module is installed
+if (-not(Get-Module -ListAvailable -Name powershell-yaml)) {
   try {
-    Install-Module powershell-yaml
+    Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.201 -Force
+    Install-Module -Name powershell-yaml -Force -Repository PSGallery -Scope CurrentUser
   } catch {
-  #Pass the exception 
-  throw [System.Net.WebException]::new("Error retrieving powershell module: powershell-yaml. Check that you have installed the Windows Package Manager modules correctly.", $_.Exception)
-  #bugbug is this good enough?
+    # If there was an exception while installing, pass it as an InternalException for further debugging
+    throw [UnmetDependencyException]::new("'powershell-yaml' was not installed successfully", $_.Exception)
+  } finally {
+    # Check to be sure it acutally installed
+    if (-not(Get-Module -ListAvailable -Name powershell-yaml)) {
+      throw [UnmetDependencyException]::new("`powershell-yaml` was not found. Check that you have installed the module correctly.")
+    }
   }
 }
 
 [System.Collections.ArrayList]$finalPackages = @()
-$configurationVersion = "0.2.0" 
+$configurationVersion = '0.2.0'
+$Utf8NoBomEncoding = New-Object System.Text.UTF8Encoding $False
+$DSCHeader = "# yaml-language-server: `$schema=https://aka.ms/configuration-dsc-schema/$($configurationVersion)"
 
-$continue = $true
-do
-{
-  $appId = Read-Host "What is the Winget ID, or name of the package you want to add to the configuration file?"
-  $findResult = Find-WinGetPackage $appId
+do {
+  $findResult = Find-WinGetPackage $(Read-Host 'What is the Winget ID, or name of the package you want to add to the configuration file?')
   
-  if ($findResult.count -ne 0)
-  {
-    $Index=1
-    foreach ($package in $findResult)
-    {
-        $packageDetails = "[$($Index)] $($package.Name) | $($package.Id) | $($package.Version)"
-        Write-Host $packageDetails
-        $index++
-    }
+  if ($findResult.count -ne 0) {
+    # Assign an index to each package
+    $findResult | ForEach-Object { $i = 1 } { Add-Member -InputObject $_ -NotePropertyName Index -NotePropertyValue $i; $i++ }
```

Comment:
IIRC; It has to be done here so that it works with the ForEach-Object

Reply:
Interesting; Will check

Reply:
After looking into this, it really doesn't matter where it go because the first script block is the `-Before` script, which is executed once when `ForEach-Object` begins processing items from the pipeline. 

The reason the PowerShell extension is complaining is because it *thinks* there is separate scopes between the `-Begin` and `-Process` section of the `ForEach-Object` and treats the `-Begin` scriptBlock as its own local scope. This may be a bug in the extension, but I'm not certain. I found that explicitly setting the scope of the initialization to be `script` fixes the module's complaint.   

While I could have moved the initialization up a line, I elected to keep it as part of the `-Begin` and explicitly set the scope to better represent the correlation between the variable and the usage since it is only used in the `ForEach-Object`.